### PR TITLE
VS: ArrayList

### PR DESCRIPTION
**Pull request by hsp-vahe**

### CODE REVIEW

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	/**
+	 * Increasing list size.
+	 */
+	private void increasListSize() {
+		array = Arrays.copyOf(array, array.length * 2);
+	}
+	
+	@Override
+	public String get(int index) {
+			validateIndex(index);
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: Format your code before giving the branch for review...

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	 */
+	public MyArrayList() {
+		this.array = new String[DEFOULT_SIZE];
+		this.lastIndex = -1;
+	}
+	
+	/**
+	 * Is list full or not.
+	 */
+	private  boolean isFull() {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: Private methods must come AFTER public methods

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	private boolean isIndexOutOfRange(int index) {
+		if (index < 0 || index > lastIndex) {
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean add(String elem) {
+		if (isFull()) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: dry violation

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/CollectionsTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Testing class for collections.
+ * @author User
+ *
+ */
+public class CollectionsTest {
+
+	public static void main(String[] args) {
+		MyArrayList list = new MyArrayList();
```
[the remark this answers is not in the document]
Comment (hsp-harut) [reply src/main/java/am/hsp/apps/oop/dataStructure/collections/CollectionsTest.java]: Also you didn't tested ALL the methods you have in the MyList interface

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		validateIndex(index);
+		moveToLeft(index);
+		--lastIndex;
+	}
+	
+	@Override
+	public void addAll(int index, String[] list) {
+		validateIndex(index);
+		for (String string : list) {
+			add(index++, string);
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: no... so in each iteration you are doing whatever written in the "add" method. And did you check what that method actually do?

You are checking in each iteration whether your internal array is full and you need to increase it or not.
You already know the count of the new elements (array length) that should be added into list, so you can check whether is there enough size in your internal array to add into it these number of new elements.

So in each iteration you are moving to right all the elements again and again? Of course, this is wrong.
You can move the elements to right by specifying the step. So if I will call moveToRightByStep(index, step) then all the elements starting from the index will be moved to the right by given step.

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		}
+	}
+	
+	@Override
+	public void clear() {
+		array = null;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: So did you try to add any element in to the list after calling this method?

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	
+	@Override
+	public int size() {
+		return lastIndex + 1;
+	}
+	
+	@Override
+	public boolean isEmpty() {
+		if (lastIndex < 0) {
+			return true;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: bad

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+		}
+		return false;
+	}
+	
+	@Override
+	public boolean contains(String elem) {
+		if (indexOf(elem) < 0) {
+			return false;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: very bad

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+		}
+		return true;
+	}
+
+	@Override
+	public int indexOf(String elem) {
+		for (int i = 0; i <= lastIndex; i++) {
+			if (elem != null && elem.equals(array[i])) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: Why you are checking element is null or not in each iteration? Is there a chance that the element could be updated after any iteration?

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	
+	/**
+	 * Increasing list size.
+	 */
+	private void increasListSize() {
+		array = Arrays.copyOf(array, array.length * 2);
+	}
+	
+	@Override
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: Generate the comments for the Overriden methods:
/* and the hit the "Enter"

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyList.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	 * Get index of the elem.
+	 */
+	int indexOf(String elem);
+	
+	/**
+	 * Set element.
+	 */
+	void set(int index, String elem);
+
+
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/collections/MyList.java]: Remove unneeded new lines... Make this everywhere

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    
+	/**
+	 * Return true if index is in range,
+	 * otherwise return false.
+	 * @param index
+	 * @return
+	 */
+	private boolean isIndexOutOfRange(int index) {
+		if (index < 0 || index > lastIndex) {
+			return true;
```
[the remark this answers is not in the document]
Comment (hsp-vahe) [reply src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: cant find the mistake

File: src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java (hunk 1 above)
Comment (hsp-vahe) [reply src/main/java/am/hsp/apps/oop/dataStructure/collections/MyArrayList.java]: what you mean saying format?

File: src/main/java/am/hsp/apps/oop/dataStructure/list/ListTest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Testing class for collections.
+ * @author User
+ *
+ */
+public class ListTest {
+
+	public static void main(String[] args) {
+		MyArrayList list = new MyArrayList();
+		list.add("1");
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/ListTest.java]: image

Vahe jan now better, but still nothing is clear.
For example your message says "MyArrayList after adding elements in the given index MyArrayList" and then you show the list elements. But it is NOT CLEAR what is the index that you have added the new element, and what is this new element that has been added. Now the one to understand whether your program runs correctly or not, need to check your code and see where you have added the element exactly and then compare it with your final result of the list.
So if your message would be "MyArrayList after adding 'four' at index '3': " then everything will be clear to user who runs your program...
Please update all the messages.

File: src/main/java/am/hsp/apps/oop/dataStructure/list/listException/IndexOutOfRangeException.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package main.java.am.hsp.apps.oop.dataStructure.list.listException;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/listException/IndexOutOfRangeException.java]: not a good package name. There is no need to indicate the "list" in the package name "listException". This package is already in the "list" package, so it is clear that whatever package you will have in it, that will correspond to the "list".
So the package name must be just "exception"

File: src/main/java/am/hsp/apps/oop/dataStructure/list/listException/IndexOutOfRangeException.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Exception class for checking index validation.
+ * @author User
+ *
+ */
+public class IndexOutOfRangeException extends RuntimeException {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/listException/IndexOutOfRangeException.java]: If you will check your code, you will see that the yellow warning is shown near this class name. 
Please autogenerate serial version id for the class. So all the classes which implement "Serializable" marker interface, must have their own serial version id.
Click on the warning icon, and select "Add generated Serial version ID"

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+
+/**
+ * Contains required fields and implementation of MyList.
+ * @author User
+ *
+ */
+public class MyArrayList implements MyList {
+	
+	private int lastIndex;
+	private String[] container;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: Why? Why just not "elements"? You store here the Elements of your list yes??

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+		this.lastIndex = -1;
+	}
+    
+	/*
+	 * (non-Javadoc)
+	 * @see main.java.am.hsp.apps.oop.dataStructure.list.MyList#get(int)
+	 */
+	@Override
+	public String get(int index) {
+			validateIndex(index);
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: You didn't fix my comment.

image

I mean that "v" of your "validateIndex" is started under the "e" of the "get".

Each new line should be just one tab righter than its enclosing block :)

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	
+	/*
+	 * (non-Javadoc)
+	 * @see main.java.am.hsp.apps.oop.dataStructure.list.MyList#addAll(int, java.lang.String[])
+	 */
+	@Override
+	public void addAll(int index, String[] list) {
+		validateIndex(index);
+		while (container.length - size() <= list.length) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: No. Don't you see that your code doesn't look good? It is of course Non Cohesive...
1. You are checking whether your list is full for the new list to be added into it.
2. You are increasing internal elements container when it is full
3. You are copying the elements from the "list" argument, into current list elements starting from given index (copy(int indexFrom, list))

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+		return indexOf(elem) >= 0;
+	}
+	
+	/*
+	 * (non-Javadoc)
+	 * @see main.java.am.hsp.apps.oop.dataStructure.list.MyList#indexOf(java.lang.String)
+	 */
+	@Override
+	public int indexOf(String elem) {
+		for (int i = 0; i <= lastIndex; i++) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: And what if the "elem" is null? So in that case you will have nullpointerexception? Is that ok?

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+			if (elem.equals(container[i])) {
+				return i;
+			}
+		}	
+		return -1;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: Remove Magic Numbers.

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+		container[index] = elem;
+	}
+	
+    /*
+     * (non-Javadoc)
+     * @see java.lang.Object#toString()
+     */
+	@Override
+	public String toString() {
+		String[] printableList = new String[lastIndex+1];
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: Why? why are you creating a new array here???

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+		for (int i = 0; i <= lastIndex; i++) {
+			printableList[i] = container[i];
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: Can't you just Syso the current element ? and that is all

File: src/main/java/am/hsp/apps/oop/dataStructure/list/exception/IndexOutOfRangeException.java [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+ * Exception class for checking index validation.
+ * @author User
+ *
+ */
+public class IndexOutOfRangeException extends RuntimeException {
+    
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 1L;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/exception/IndexOutOfRangeException.java]: Wrong. I told you to use **"Add generated Serial version ID"** option and not have the 1L...

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 24]
```diff
@@ -0,0 +1,271 @@
...
+		return " [" + Arrays.toString(elements) + "]";
+	}
+	
+	/**
+	 * Print list.
+	 */
+	public void printList() {
+		System.out.print(" [");
+		for (int i = 0; i <= lastIndex; i++) {
+			System.out.print(" " + elements[i] + ",");
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: You always print the comma ","? Even after the last element? That is wrong.

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+		validateIndex(index);
+		elements[index] = elem;
+	}
+	
+    /*
+     * (non-Javadoc)
+     * @see java.lang.Object#toString()
+     */
+	@Override
+	public String toString() {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: So if you have this method, then why you need the "printList" method at all? Don't they do the same?

File: src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+	public void set(int index, String elem) {
+		validateIndex(index);
+		elements[index] = elem;
+	}
+
+	/**
+	 * Print list.
+	 */
+	public void printList() {
+		System.out.print(" [" + elements[0]);
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/oop/dataStructure/list/arrayList/MyArrayList.java]: So what if you list is empty? In that case what you would print here?

**Pull request merged**